internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,333 @@
+			}
+		}
+	}
+	if selector.MatchResources.MachineDeploymentClass != nil {
+		for _, name := range selector.MatchResources.MachineDeploymentClass.Names {
+			for _, md := range class.Spec.Workers.MachineDeployments {
+				if md.Class == name {
+					if selectorMatchTemplate(selector, md.Template.Infrastructure.Ref) {
+						return nil
+					}
+					if selectorMatchTemplate(selector, md.Template.Bootstrap.Ref) {
+						return nil
+					}
+				}
+			}
+		}
+	}
+	// if the code has not returned at this point there is no matching template in the ClusterClass. Return an error.
+	return append(allErrs,
+		field.Invalid(
+			path,
+			selector,
+			"selector did not match any template in the ClusterClass",
+		))
+}
+
+// selectorMatchTemplate returns true if APIVersion and Kind for the given selector match the reference.
+func selectorMatchTemplate(selector clusterv1.PatchSelector, reference *corev1.ObjectReference) bool {
+	return selector.Kind == reference.Kind && selector.APIVersion == reference.APIVersion
+}
+
+var validOps = sets.NewString("add", "replace", "remove")
+
+func validateJSONPatches(jsonPatches []clusterv1.JSONPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	variableSet := getClusterClassVariablesMap(variables)
+
+	for i, jsonPatch := range jsonPatches {
+		if !validOps.Has(jsonPatch.Op) {
+			allErrs = append(allErrs,
+				field.NotSupported(
+					path.Index(i).Child("op"),
+					jsonPatch,
+					validOps.List(),
+				))
+		}
+
+		if !strings.HasPrefix(jsonPatch.Path, "/spec/") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch path must start with \"/spec/\".",
+				))
+		}
+
+		// return early if the jsonPatch does not have "replace" or "add" op. Additional validation is not needed.
+		if jsonPatch.Op != "add" && jsonPatch.Op != "replace" {
+			return allErrs
+		}
+
+		if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch must define one of value or valueFrom.",
+				))
+		}
+
+		if jsonPatch.Value != nil && jsonPatch.ValueFrom != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch can not define both value and valueFrom.",
+				))
+		}
+
+		// Attempt to marshal the JSON to discover if it  is valid.
+		if jsonPatch.Value != nil {
+			if err := json.Unmarshal(jsonPatch.Value.Raw, &struct{}{}); err != nil {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Index(i),
+						jsonPatch.Value,
+						"jsonPatch Value is invalid JSON.",
+					))
+			}
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template == nil && jsonPatch.ValueFrom.Variable == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i).Child("valueFrom"),
+					jsonPatch,
+					"valueFrom must set either \"/template or variable/\".",
```

Comment:
```suggestion
					"valueFrom must set either template or variable.",
```
The escaping looks a bit awkward (especially the additional leading/trailing slash). I think we should either use the field names consistently in plain text or quote the individual field names (would also apply to all occurrences of `value` and `valueFrom`